File: java/kindergarten-garden/src/main/java/KindergartenGarden.java
import java.util.Arrays;
import java.util.List;
import java.util.Vector;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class KindergartenGarden {

  private static final int CUPS_PER_ROW = 2;
  private static final String[] DEFAULT_STUDENTS = { "Alice", "Bob", "Charlie", "David", "Eve", "Fred", "Ginny", "Harriet", "Ileana","Joseph", "Kincaid", "Larry" };      
  private final Vector<Vector<Character>> rows = new Vector<Vector<Character>>();
  private final List<String> students;

  public KindergartenGarden(String plants) { this(plants, DEFAULT_STUDENTS); }  
  public KindergartenGarden(String plants, String[] studentArray) {           
    this.students = Arrays.asList(studentArray);    
    this.students.sort(String::compareTo);
    Arrays.stream(plants.split("\\r?\\n"))
     .forEach(row -> this.rows.addElement(new Vector<Character>(
         row.chars().mapToObj(c -> (char)c).collect(Collectors.toList()))));
  }

  public List<Plant> getPlantsOfStudent(String student) {             
    int studentIndex = students.indexOf(student);        
    return IntStream.range(0, rows.size())
      .mapToObj(i -> rows.get(i).stream().skip(studentIndex * CUPS_PER_ROW).limit(CUPS_PER_ROW))
      .flatMap(c -> c.map(Plant::getPlant))
      .collect(Collectors.toList());
  }
}

File: csharp/kindergarten-garden/KindergartenGarden.cs
using System;
using System.Linq;
using System.Collections.Generic;

public class KindergartenGarden
{
  private const int WINDOW_SIZE = 2;
  private static readonly string[] DEFAULT_STUDENTS = { "Alice", "Bob", "Charlie", "David", "Eve", "Fred", "Ginny", "Harriet", "Ileana", "Joseph", "Kincaid", "Larry" };
  private static readonly Dictionary<char, Plant> PLANT_MAPPING = new Dictionary<char, Plant>
  {
    { 'V', Plant.Violets },
    { 'R', Plant.Radishes },
    { 'C', Plant.Clover },
    { 'G', Plant.Grass }
  };

  private readonly Dictionary<string, List<Plant>> _plantsPerStudent;

  public KindergartenGarden(string diagram) : this(diagram, DEFAULT_STUDENTS) { }
  public KindergartenGarden(string diagram, IEnumerable<string> students)
  {
    _plantsPerStudent = students.ToDictionary(x => x, x => new List<Plant>());
    var rows = diagram.Split(new[] { "\n" }, StringSplitOptions.RemoveEmptyEntries);
    foreach (var row in rows)
    {
      var offset = 0;
      foreach (var student in _plantsPerStudent.Keys)
      {
        _plantsPerStudent[student].AddRange(row.Skip(offset).Take(WINDOW_SIZE).Select(x => PLANT_MAPPING[x]));
        offset += WINDOW_SIZE;
      }
    }
  }

  public IEnumerable<Plant> Plants(string student)
  {
    return _plantsPerStudent[student].AsReadOnly();
  }
}

public enum Plant
{
  Violets,
  Radishes,
  Clover,
  Grass
}